drivers/ltc4150/ltc4150.c

```diff
@@ -0,0 +1,272 @@
+ */
+
+/**
+ * @ingroup     drivers_ltc4150
+ * @brief       Driver for the Linear Tech LTC4150 Coulomb Counter
+ *              (a.k.a. battery gauge sensor or power consumption sensor)
+ * @{
+ *
+ * @file
+ * @brief       LTC4150 Device Driver
+ * @author      Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ *
+ * @}
+ */
+#include <errno.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "irq.h"
+#include "ltc4150.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static void update_data(ltc4150_dev_t *dev)
+{
+    timex_t now;
+
+    xtimer_now_timex(&now);
+    // For each second passed since last update:
+    while (dev->last_update_sec < now.seconds) {
+        // Update running avg for seconds
+#ifdef MODULE_LTC4150_BATTERY
+        dev->charged.last_min += dev->charged.counter;
+        dev->charged.last_min -= dev->charged.sec[dev->position];
+        dev->charged.sec[dev->position] = dev->charged.counter;
+        dev->charged.counter = 0;
+#endif
+
+        dev->discharged.last_min += dev->discharged.counter;
+        dev->discharged.last_min -= dev->discharged.sec[dev->position];
+        dev->discharged.sec[dev->position] = dev->discharged.counter;
+        dev->discharged.counter = 0;
+
+        dev->position++;
+
+        // Every 60 second update running avg for minutes
+        if (dev->position >= 60) {
+            dev->position = 0;
+        }
+
+        dev->last_update_sec++;
+    }
+}
+
+static void pulse_cb(void *_dev)
+{
+    ltc4150_dev_t *dev = _dev;
+
+    update_data(dev);
+#ifdef MODULE_LTC4150_BATTERY
+    if ((dev->polarity == GPIO_UNDEF) || (!gpio_read(dev->polarity))) {
+        dev->discharged.counter++;
+        dev->discharged.total++;
+    }
+    else {
+        dev->charged.counter++;
+        dev->charged.total++;
+    }
+#else
+    dev->discharged.counter++;
+    dev->discharged.total++;
+#endif
+}
+
+int ltc4150_init(ltc4150_dev_t *dev, const ltc4150_params_t *params)
+{
+    timex_t now;
+
+    if (!dev || !params) {
+        return -EINVAL;
+    }
+
+    memset(dev, 0x00, sizeof(ltc4150_dev_t));
+    dev->interrupt = params->interrupt;
+#ifdef MODULE_LTC4150_BATTERY
+    dev->polarity = params->polarity;
+#endif
+    dev->shutdown = params->shutdown;
+    dev->pulses_per_ah = params->pulses_per_ah;
+
+    if (dev->shutdown != GPIO_UNDEF) {
+        if (gpio_init(dev->shutdown, GPIO_OUT)) {
+            DEBUG("[ltc4150] Failed to initialize shutdown pin");
+            return -EIO;
+        }
+        gpio_set(dev->shutdown);
+    }
+
+#ifdef MODULE_LTC4150_BATTERY
+    if (dev->polarity != GPIO_UNDEF) {
+        if (gpio_init(dev->polarity, GPIO_IN)) {
+            DEBUG("[ltc4150] Failed to initialize polarity pin");
+            return -EIO;
+        }
+    }
+#endif
+
+    xtimer_now_timex(&now);
+    dev->last_update_sec = now.seconds;
+    dev->start_sec = now.seconds;
+
+    if (gpio_init_int(dev->interrupt, GPIO_IN, GPIO_FALLING, pulse_cb, dev)) {
```

Comment:
It might make sense to allow configuring it using `ltc4150_params_t`. But both the breakout board and the MSBA2 have external pull up resistors, so in that cases it seems better to rely on the external pull up resistor where it is present.

Reply:
I'm fine with that